### PR TITLE
feat(integer): improve mul and scalar mul

### PR DESCRIPTION
**Pull request by tmontaigu**
This improves the mul and scalar_mul algorithms
to be faster

The improvement is made within the code
that was responsible for summing up all
the terms by making better use of carries
and avoiding uncessary propagations.

The scalar mul forwards the call to a right shift
when the scalar is a power of two as it just cost one PBS so it will always be faster.

For 64-bits, target-cpu=native + avx512:
- mul before: 3.4s
- mul after: 900ms

### CODE REVIEW

File: tfhe/src/integer/server_key/radix_parallel/mul.rs [hunk 1]
```diff
@@ -273,6 +273,148 @@ impl ServerKey {
...
+        let num_blocks = lhs.blocks.len();
+
+        let num_bits_in_carry = self.key.carry_modulus.0.ilog2() as u64;
+        // Among those bits of carry, we know one is already consumed by
+        // the last step of the unchecked_block_mul_parallelized
+        // for each of the term.
+        //
+        // It means we can only do num_bits_in_carry - 1 additions
+        // Which then means we can iter on chunks of num_bits_in_carry
+        // to add themp up as it will result in num_bits_in_carry - 1 additions
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/mul.rs:296]: themp -> them

File: tfhe/src/integer/server_key/radix_parallel/mul.rs [hunk 2]
```diff
@@ -273,6 +273,148 @@ impl ServerKey {
...
+        let chunk_size = num_bits_in_carry as usize;
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/mul.rs:297]: so for 2_2 chunk_size = 2, it's a binary tree reduction ? Did you look at other parameter sets by curiosity ?

does this algorithms work with 1_1 params ? 🤔

File: tfhe/src/integer/server_key/radix_parallel/mul.rs [hunk 3]
```diff
@@ -273,6 +273,148 @@ impl ServerKey {
...
+
+        // For the last chunk, we want to finish it off
+        // using an addition that does not leave the resulting ciphertext with
+        // non empty carries.
+        //
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/mul.rs:302]: `we want to use an addition that leaves the resulting ciphertext with empty carries`

seems more readable if I got what you wanted to say

File: tfhe/src/integer/server_key/radix_parallel/mul.rs [hunk 4]
```diff
@@ -273,6 +273,148 @@ impl ServerKey {
...
+                for (ct_left_i, ct_right_i) in message_blocks
+                    [first_block_where_addition_happenned + 1..]
+                    .iter_mut()
+                    .zip(carry_blocks.iter())
+                {
+                    self.key.unchecked_add_assign(ct_left_i, ct_right_i);
+                }
+            });
+
+            // temrs is organized like so:
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/mul.rs:361]: tmers -> terms

File: tfhe/src/integer/server_key/radix_parallel/mul.rs (hunk 3 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/mul.rs:313]: yep my phrasing was a bit confusing

**Comment by tmontaigu** [comment]
@slab-ci cpu_test
@slab-ci cpu_integer_test

File: tfhe/src/integer/server_key/radix_parallel/mul.rs (hunk 2 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/mul.rs]: fixed for 1_1

**Comment by tmontaigu** [comment]
@slab-ci cpu_integer_test

File: tfhe/src/integer/server_key/radix_parallel/mul.rs (hunk 3 above)
Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/mul.rs:313]: looks like this was not updated, let's keep it for a next PR

**Review by IceTDrinker — APPROVED** [review]
Good for me once tests are green

**Comment by IceTDrinker** [comment]
thanks a lot, looks good to merge once tests are green !

**Pull request merged**